src/bitcoin/poller/mod.rs

```diff
@@ -1,60 +1,131 @@
+        looper::poll(&self.bit, &self.db, &self.secp, &self.descs);
     }
 
-    #[cfg(test)]
-    pub fn test_stop(&mut self) {
-        self.shutdown.store(true, atomic::Ordering::Relaxed);
+    /// Continuously update our state from the Bitcoin backend.
+    /// - `poll_interval`: how frequently to perform an update.
+    /// - `shutdown`: set to true to stop continuously updating and make this function return.
+    ///
+    /// Typically this would run for the whole duration of the program in a thread, and the main
+    /// thread would set the `shutdown` atomic to `true` when shutting down.
+    pub fn poll_forever(
+        &self,
+        poll_interval: time::Duration,
+        receiver: mpsc::Receiver<PollerMessage>,
+    ) {
+        let mut last_poll = None;
+        let mut synced = false;
+
+        loop {
+            // How long to wait before the next poll.
+            let time_before_poll = if let Some(last_poll) = last_poll {
+                let time_since_poll = time::Instant::now().duration_since(last_poll);
+                // Until we are synced we poll less often to avoid harassing bitcoind and impeding
+                // the sync. As a function since it's mocked for the tests.
+                let poll_interval = if synced {
+                    poll_interval
+                } else {
+                    looper::sync_poll_interval()
+                };
+                poll_interval.saturating_sub(time_since_poll)
+            } else {
+                // Don't wait before doing the first poll.
+                time::Duration::ZERO
+            };
+
+            // Wait for the duration of the interval between polls, but listen to messages in the
+            // meantime.
+            match receiver.recv_timeout(time_before_poll) {
+                Ok(PollerMessage::Shutdown) => {
+                    log::info!("Bitcoin poller was told to shut down.");
+                    return;
+                }
+                Ok(PollerMessage::PollNow(sender)) => {
+                    // We've been asked to poll, don't wait any further and signal completion to
+                    // the caller.
+                    self.poll(&mut last_poll);
+                    if let Err(e) = sender.send(()) {
+                        log::error!("Error sending immediate poll completion signal: {}.", e);
+                    }
+                    continue;
+                }
+                Err(mpsc::RecvTimeoutError::Timeout) => {
+                    // It's been long enough since the last poll.
+                }
+                Err(mpsc::RecvTimeoutError::Disconnected) => {
+                    log::error!("Bitcoin poller communication channel got disconnected. Exiting.");
+                    return;
+                }
+            }
+
+            // Don't poll until the Bitcoin backend is fully synced.
+            if !synced {
+                let progress = self.bit.sync_progress();
+                log::info!(
```

Comment:
Maybe pass this log in debug, in liana-gui console we get a lot of:
```

  2024-03-21T10:36:36.628035Z  INFO liana::bitcoin::poller:116: Block chain synchronization progress: 99.90% (187699 blocks / 187728 headers)

  2024-03-21T10:36:36.664726Z  INFO liana::bitcoin::poller:116: Block chain synchronization progress: 99.90% (187705 blocks / 187728 headers)

  2024-03-21T10:36:36.704515Z  INFO liana::bitcoin::poller:116: Block chain synchronization progress: 99.90% (187711 blocks / 187728 headers)

  2024-03-21T10:36:36.736193Z  INFO liana::bitcoin::poller:116: Block chain synchronization progress: 99.90% (187711 blocks / 187728 headers)

  2024-03-21T10:36:36.770596Z  INFO liana::bitcoin::poller:116: Block chain synchronization progress: 99.90% (187717 blocks / 187728 headers)

  2024-03-21T10:36:36.807980Z  INFO liana::bitcoin::poller:116: Block chain synchronization progress: 99.90% (187723 blocks / 187728 headers)

  2024-03-21T10:36:36.845011Z  INFO liana::bitcoin::poller:116: Block chain synchronization progress: 100.00% (187728 blocks / 187728 headers)

```
While the sync

Reply:
I tried not to change behaviour in this PR.

Reply:
Hmm but behaviour did change, looks like recv_timeout doesn't block, or not for long enough.